calc/calc_mamba_flops.py

```diff
@@ -59,160 +66,187 @@ def config_parser():
+        total_flops = embedding_flops + (mamba1_flops * args.num_mamba_layers)
+        total_mamba1_flops = mamba1_flops * args.num_mamba_layers
+        
         # if MoE layers add these in
         if args.num_moe_layers > 0:
-            ffn_flops = iter_factor * args.tokens  * 4 * args.ffn_hidden_size * args.num_moe_layers * args.hidden_size
+            ffn_flops = iter_factor * args.tokens * 4 * args.ffn_hidden_size * args.num_moe_layers * args.hidden_size
             if args.swiglu:
                 ffn_flops = 3/2 * ffn_flops
             gating_flops = iter_factor * 2 * args.tokens * args.num_experts * args.num_moe_layers
             total_flops += ffn_flops + gating_flops
             total_ffn_flops += ffn_flops
         
     else:
-        arch_list = args.mamba_moe_layers.split(" ")
-        total_flops = 0
-        total_flops += embedding_params
-        for el in arch_list:
-            if el == "r":
-                # mamba layer
-                total_flops += mamba_flops
-                total_mamba_flops += mamba_flops
-            elif el == "m":
+        total_flops = embedding_flops
+        for layer_type in args.mamba_moe_layers.split():
+            if layer_type == "r":
+                # mamba1 layer
+                total_flops += mamba1_flops
+                total_mamba1_flops += mamba1_flops
+            elif layer_type == "m":
+                # mamba2 layer
                 total_flops += mamba2_flops
                 total_mamba2_flops += mamba2_flops
-            elif el == "a":
+            elif layer_type == "a":
+                # attention layer
                 total_flops += attention_flops 
                 total_attention_flops += attention_flops
-            elif el.isnumeric():
+            elif layer_type.isnumeric():
+                # FFN layer
                 total_flops += ffn_flops
                 total_ffn_flops += ffn_flops
-            elif el == "g":
-                # zamba shared layer
+            elif layer_type == "g":
+                # shared layer case
                 original_hidden_size = args.hidden_size
                 args.hidden_size = original_hidden_size * 2
-                shared_attention_flops = compute_attention_flops(args)
+                shared_attention_flops = compute_attention_flops(args, iter_factor)
                 shared_ffn_flops = compute_ffn_flops(args)
-                total_flops += shared_attention_flops
+                total_flops += shared_attention_flops + shared_ffn_flops
                 total_attention_flops += shared_attention_flops
-                total_flops += shared_ffn_flops
-                total_ffn_flops = shared_ffn_flops
+                total_ffn_flops += shared_ffn_flops
                 args.hidden_size = original_hidden_size
                 # final downprojector matrix
                 total_flops += 4 * args.batch_size * args.sequence_length * args.hidden_size * args.hidden_size
```

Comment:
```suggestion
                total_flops += 2 * args.batch_size * args.sequence_length * args.hidden_size * args.hidden_size * iter_factor
```
this is a square matrix, the downproj happens inside attention as commented above